src/osgPlugins/dds/ReaderWriterDDS.cpp

```diff
@@ -1010,7 +1028,27 @@ osg::Image* ReadDDSFile(std::istream& _istream, bool flipDDSRead)
         // this memory will not be used but it will not cause leak in worst meaning of this word.
     }
 
-    osgImage->setImage(s,t,r, internalFormat, pixelFormat, dataType, imageData, osg::Image::USE_NEW_DELETE, packing);
+    if (ddsd.ddpfPixelFormat.dwFlags & DDPF_PALETTEINDEXED8)
+    {
+        // Now we need to substitute the indexed image data with full RGBA image data.
+        unsigned char* convertedData = new unsigned char [sizeWithMipmaps * 4];
+        for (unsigned int i = 0; i < sizeWithMipmaps; i++)
+        {
+            convertedData[i * 4 + 0] = palette[imageData[i] * 4 + 0];
+            convertedData[i * 4 + 1] = palette[imageData[i] * 4 + 1];
+            convertedData[i * 4 + 2] = palette[imageData[i] * 4 + 2];
+            convertedData[i * 4 + 3] = palette[imageData[i] * 4 + 3];
+        }
+        for (unsigned int i = 0; i < mipmap_offsets.size(); i++)
+            mipmap_offsets[i] *= 4;
+        internalFormat = GL_RGBA;
+        pixelFormat = GL_RGBA;
+        osgImage->setImage(s,t,r, internalFormat, pixelFormat, dataType, convertedData, osg::Image::USE_NEW_DELETE, packing);
```

Comment:
What's about `imageData` array? I suppose we should free it in this scope to avoid a memory leak.

Reply:
Yeah, corrected. I'm too used to automatic memory allocation and deallocation. :smile:

Reply:
In general I prefer readability over hypothetical code optimizations, I say hypothetical as compiler optimizer are pretty good.  It's also really important to know the roll of the code - it's it code that is on an important inner loop of rendering that could make a difference to fps, or is it an ancillary feature that rarely gets called.  

For that really performance critical it's more warrented to spend time and optimize carefully and be willing to compromise on readability for these improvements.  For the dds plugin, it's something run at start up prior to rendering, and only be a small number of users, and those time it's used during rendering is safely tucked away in a background database thread so... never a performance critical feature.

So it comes down to which version is more readable.  I'm so used to unrolling pointer management like @mp3butcher's variant that I find it just as readable as the original, if not more so.  I'm pretty old school though :-)